tests: Add unix timestamp tests for OP_CLTV and max mediantimepast tests

Christewart opened the pull request:
This PR adds a functional (p2p/block-validation) test for `OP_CHECKLOCKTIMEVERIFY` (BIP65) using unix-timestamp locktimes in test/functional/feature_cltv.py. It also adds tests around the maximum possible median-time-past (MTP) value, `UINT32_MAX`, and how that interacts with transaction finality.

The unix-timestamp portion of BIP65 already has unit-level coverage in `transaction_tests.cpp` (via `tx_valid.json`/`tx_invalid.json`), but those tests call `VerifyScript()` directly and never exercise median-time-past (MTP) or block connection. MTP-vs-nLockTime behavior itself is covered separately in `miner_tests.cpp` (lines 525-534) and `feature_csv_activation.py` (lines 340-355), but neither exercises `OP_CLTV` specifically. This test closes that gap: it builds and submits full blocks over p2p and checks that a block containing an `OP_CLTV`-locked output is only accepted once BIP113 MTP actually satisfies the requested locktime.

This PR also adds a BIP113 test to `feature_csv_activation.py` showing that MTP can never advance past `UINT32_MAX` - 1, so an output with a plain nLockTime of `UINT32_MAX` can never be confirmed. This is a general BIP113 property independent of `OP_CLTV`.

Thanks to @mabu44 and @sedited for the review so far.

mabu44 reviewed (commented):
The Unix-timestamp portion of BIP65 is currently tested in the c++ unit tests in "transactions_tests.cpp". Test data includes transactions that have OP_CLTV with Unix timestamps (the files are /src/test/data/tx_valid.json and tx_invalid.json).
Does your test cover scenarios that are not covered in the existing tests?

Christewart commented:
Hi @mabu44 . Thanks for taking a look at this.

>  Do your test cover scenarios that are not covered in the existing tests?

Yes I believe so. I took a look at the values in `tx_valid.json` and `tx_invalid.json`. Here is a table

| OP_CLTV parameter | Transaction locktime field | Valid |
|-------------------|----------------------------|--------|
| 0                 | 0                          | true   |
| 499999999         | 499999999                  | true   |
| 0                 | 499999999                  | true   |
| 500000000         | 500000000                  | true   |
| 4294967295        | 4294967295                 | true   |
| 500000000         | 4294967295                 | true   |
| 0                 | 0                          | true   |
| -11               | 500000000                  | true   |
| -19               | 4294967294                 | true   |
| 0                 | 0                          | true   |
| 1                 | 0                          | false  |
| 499999999         | 499999998                  | false  |
| 500000001         | 500000000                  | false  |
| 4294967295        | 4294967294                 | false  |
| -1                | 0                          | false  |
| -1                | 500000000                  | false  |
| 0                 | 0                          | false  |
| 0                 | 0                          | false  |
| 0                 | 500000000                  | false  |
| 499999999         | 500000000                  | false  |
| 500000000         | 0                          | false  |
| 500000000         | 499999999                  | false  |
| 4294967296        | 4294967295                 | false  |
| 2147483648        | 2147483647                 | false  |
| 0                 | 0                          | false  |

This test also directly tests the median-time-past logic. This is not tested in `tx_{valid,invalid}.json` as we just [directly call `VerifyScript()`](https://github.com/bitcoin/bitcoin/blob/cfe025ff0e1dcf984179a5a9b01dc8b865d95409/src/test/transaction_tests.cpp#L129) for individual transactions in the test files.

Interestingly, i think there may be one test case overlap with `tx_valid.json`. While this transaction could be valid with relay policy I don't believe a transaction with these locktimes would be allowed to be included in a block? This is because median-time-past logic for a block with this transaction could be not be satisfied as per my last test case [here](https://github.com/Christewart/bitcoin/blob/69b6bb9438095f84ce90aecfc4e76b0ff21cc2b8/test/functional/feature_cltv.py#L256).

>| 4294967295        | 4294967295                 | true   |

Some more eyes on this could be good as I may be making a mistake.

Christewart commented on test/functional/feature_cltv.py:
```diff
@@ -199,6 +201,67 @@ def run_test(self):
+
+        # now our median-past-time should allow for a transaction with
+        # a locktime of +1 hour to be confirmed
+        spendtx = wallet.create_self_transfer()['tx']
```
I believe so because the previous one was mined in a valid block here: https://github.com/Christewart/bitcoin/blob/69b6bb9438095f84ce90aecfc4e76b0ff21cc2b8/test/functional/feature_cltv.py#L213

mabu44 reviewed (commented):
So this is an integration test of:
- *Transaction validity based on input script and transaction locktime*: currently tested in [transactions_tests.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/test/transaction_tests.cpp);
- *Transaction validity based on median block time*: currently tested in [miner_tests.cpp (lines 525-534)](https://github.com/bitcoin/bitcoin/blob/b8cefeb221490868d62b7a0695f8b5ea392d3654/src/test/miner_tests.cpp#L525-L534) and [feature_csv_activation.py (lines 340-355)](https://github.com/bitcoin/bitcoin/blob/b8cefeb221490868d62b7a0695f8b5ea392d3654/test/functional/feature_csv_activation.py#L340-L355)

with some more edge cases relative to the year 2106. Right?

> Interestingly, i think there may be one test case overlap with tx_valid.json. While this transaction could be valid with relay policy I don't believe a transaction with these locktimes would be allowed to be included in a block? This is because median-time-past logic for a block with this transaction could be not be satisfied as per my last test case [here](https://github.com/Christewart/bitcoin/blob/69b6bb9438095f84ce90aecfc4e76b0ff21cc2b8/test/functional/feature_cltv.py#L256).

>> | 4294967295 | 4294967295 | true |

I think that such a transaction will become relay-valid in 2106, but, unless consensus changes, it will never be mined because no blocks will be produced to mine it.

Christewart commented:
Thank you again for the careful review. I've integrated your code review in 024c3bf

> So this is an integration test of:
> 
>     * _Transaction validity based on input script and transaction locktime_: currently tested in [transactions_tests.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/test/transaction_tests.cpp);
> 
>     * _Transaction validity based on median block time_: currently tested in [miner_tests.cpp (lines 525-534)](https://github.com/bitcoin/bitcoin/blob/b8cefeb221490868d62b7a0695f8b5ea392d3654/src/test/miner_tests.cpp#L525-L534) and [feature_csv_activation.py (lines 340-355)](https://github.com/bitcoin/bitcoin/blob/b8cefeb221490868d62b7a0695f8b5ea392d3654/test/functional/feature_csv_activation.py#L340-L355)
> 
> 
> with some more edge cases relative to the year 2106. Right?

Yes.

> > Interestingly, i think there may be one test case overlap with tx_valid.json. While this transaction could be valid with relay policy I don't believe a transaction with these locktimes would be allowed to be included in a block? This is because median-time-past logic for a block with this transaction could be not be satisfied as per my last test case [here](https://github.com/Christewart/bitcoin/blob/69b6bb9438095f84ce90aecfc4e76b0ff21cc2b8/test/functional/feature_cltv.py#L256).
> 
> > > | 4294967295 | 4294967295 | true |
> 
> I think that such a transaction will become relay-valid in 2106, but, unless consensus changes, it will never be mined because no blocks will be produced to mine it.

That is my understanding as well :+1:

fanquake commented:
cc @petertodd

fanquake commented:
@Christewart any chance you want to follow up to the above?

Christewart commented:
Hi @sedited 

> it looks like we are already testing unix timestamp values in the locktime extensively.

`feature_csv_activation.py` tests relative lock times and the `OP_CHECKSEQUENCEVERIFY` opcode. This test case is testing `OP_CHECKLOCKTIMEVERIFY` and absolute wall clock locktimes ("unix timestamps"). Different fields on the transaction are used for verification (`sequence` vs `locktime`). 

>Are the finality tests better suited for feature_csv_activation

In my opinion no so since this is testing absolute wall clock locktimes via a different opcode - `OP_CLTV`.

The real interesting part to me is the [setting the unix timestamp of the node to `UINT32_MAX`](https://github.com/Christewart/bitcoin/blob/cbc1095adb8f729c2cb06574144187a16f45c68a/test/functional/feature_cltv.py#L239), and then attempting to spend `OP_CLTV` locked utxos when we have reached the `UINT32_MAX` timestamp in the locktime field. 

Since we cannot satisfy the MTP requirement of BIP113 (i.e. we cannot create a block with timestamp `UINT32_MAX+1` so we are following BIP113 and monotonically increase the timestamp on the block) - the utxo can never be spent.

From a quick glance through `feature_csv_activation.py` I don't see any references to `2**32` or `UINT32_MAX` - so I doubt this edge case is tested there - although if you could link to an exact line that it was tested on that would be helpful.

In general, I think @mabu44 's comment was good:

https://github.com/bitcoin/bitcoin/pull/32229#pullrequestreview-2756380784


EDIT: 

Looking closer at `feature_csv_activation.py` it does look like it tests BIP113 logic (its unfortunate the file is named after the opcode and is testing more than the opcode). The logic testing `UINT32_MAX` could be moved over to that file, however the `OP_CLTV` test logic should stay here and I still believe there is no integration tests for OP_CLTV and unix timestamps - please link if you find them

sedited commented:
> The logic testing UINT32_MAX could be moved over to that file,

Yes, sorry for not being clearer on this before, but that is what I was suggesting.

> however the OP_CLTV test logic should stay here and I still believe there is no integration tests for OP_CLTV and unix timestamps - please link if you find them.

Thanks for linking mabu44's comment again, I was referring to the same tests too. Maybe mention them in the PR descritpion? 

> The real interesting part to me is the [setting the unix timestamp of the node to UINT32_MAX](https://github.com/Christewart/bitcoin/blob/cbc1095adb8f729c2cb06574144187a16f45c68a/test/functional/feature_cltv.py#L239), and then attempting to spend OP_CLTV locked utxos when we have reached the UINT32_MAX timestamp in the locktime field.

I guess once that MTP is reached, we'll have to change some part of the calculation anyway, but I'm fine with having a test for documenting this.

Concept ACK

Christewart commented:
Hi @sedited , I pushed a new commit: 533d0c608312beaeb47bb5373169a17f9c756814
 
Here are the things that i've changed since you've last reviewed

   - Rewrote the PR description to clarify what this actually tests and to credit the existing coverage you and @mabu44 pointed to (see above).
   - Split the `UINT32_MAX` / median-time-past edge case out of `feature_cltv.py` and into `feature_csv_activation.py`, since that behavior is a general BIP113 property and not specific to `OP_CLTV` (per our earlier discussion).
   - Rebased onto latest master.
   - The rebase pulled in #35089 (`create_block`'s `height=` kwarg replacing the old `create_coinbase(...)` calling convention), which broke a few call sites this PR touches. Fixed those up to match the new convention.


Let me know if this addresses your concerns or if there's anything else you'd like changed.

sedited commented on test/functional/feature_cltv.py:
```diff
@@ -199,6 +201,40 @@ def run_test(self):
+        locktime = current_time + (60 * 60) # 1 hour
+        spendtx = wallet.create_self_transfer()['tx']
+        self.log.info("Test that one block mined with correct wall clock time does not satisfy median-time-past")
+        cltv_validate(spendtx,locktime)
```
Nit: Missing space (and for the other `cltv_validate` call too).

sedited commented on test/functional/feature_cltv.py:
```diff
@@ -199,6 +201,40 @@ def run_test(self):
+
+        # now our median-past-time should allow for a transaction with
+        # a locktime of +1 hour to be confirmed
+        cltv_validate(spendtx,locktime)
```
If I remove this line, the test still passes. Is it required? Doesn't this also just append another scriptsig to the existing one?

sedited commented on test/functional/feature_cltv.py:
```diff
@@ -199,6 +201,40 @@ def run_test(self):
+        block2.solve()
+        peer.send_and_ping(msg_block(block2))
+        # expect failure as we haven't satisfied wall clock time yet
+        assert_not_equal(int(self.nodes[0].getbestblockhash(), 16), block2.hash_int)
```
How about:
```
        # expect failure as we haven't satisfied wall clock time yet
        with self.nodes[0].assert_debug_log(expected_msgs=['bad-txns-nonfinal']):
            peer.send_and_ping(msg_block(block2))
            assert_not_equal(int(self.nodes[0].getbestblockhash(), 16), block2.hash_int)
```
As done further above in the file to capture the rejection reason?

sedited commented on test/functional/feature_csv_activation.py:
```diff
@@ -474,6 +479,46 @@ def run_test(self):
+            self.miniwallet.sign_tx(tx)
+            return tx
+
+        # Mine 6 blocks at UINT32_MAX - 1 so that MTP (the median of the last 11 blocks) becomes UINT32_MAX - 1
+        for _ in range(6):
+            block = create_block(self.tip, height=self.tipheight + 1, ntime=UINT32_MAX - 1)
+            block.solve()
+            self.send_blocks([block])
+            self.tip = block.hash_int
+            self.tipheight += 1
+        assert_equal(self.nodes[0].getblockheader(self.nodes[0].getbestblockhash())['mediantime'], UINT32_MAX - 1)
+
+        # UINT32_MAX - 2 is the highest locktime that can be included in the next block:
+        # BIP113 requires nLockTime to be strictly less than the MTP of the previous block
+        max_locktime_tx = mtp_edge_case_tx(mtp_edge_case_inputs[0], UINT32_MAX - 2)
+        block = create_block(self.tip, height=self.tipheight + 1, ntime=UINT32_MAX, txlist=[max_locktime_tx])
+        block.solve()
+        self.send_blocks([block])
+        self.tip = block.hash_int
+        self.tipheight += 1
+
+        # A transaction locked to UINT32_MAX itself is never final: MTP can never exceed
+        # UINT32_MAX, so "nLockTime < MTP" can never be satisfied for nLockTime == UINT32_MAX
+        never_final_tx = mtp_edge_case_tx(mtp_edge_case_inputs[1], UINT32_MAX)
+        block = create_block(self.tip, height=self.tipheight + 1, ntime=UINT32_MAX, txlist=[never_final_tx])
+        block.solve()
+        self.send_blocks([block], success=False, reject_reason='bad-txns-nonfinal')
```
Is this really exercising the edge case behaviour? It seems kind of obvious that for as long as the MTP isn't maxed out, the transaction will fail. From what I gather we want to show that a transaction with nLockTime will never be mined. Wouldn't it be better to mine to an MTP of UINT32_MAX, and then show that a) submitting a transaction to the mempool with nLockTime at UINT32_MAX fails, and that extending the chain further with a block containing the transaction will fail with "time-too-old".

Christewart committed:
test: add functional coverage for OP_CLTV unix-timestamp locktimes and BIP113 interaction

Add a functional (p2p/block-validation) test for OP_CHECKLOCKTIMEVERIFY
using unix-timestamp locktimes, including its interplay with BIP113
median-time-past (MTP) rules: a block containing an OP_CLTV-locked
output is only accepted once MTP satisfies the requested locktime.

Also add a BIP113 test to feature_csv_activation.py showing that MTP
can never exceed UINT32_MAX, so an output with a plain nLockTime of
UINT32_MAX can never be confirmed: BIP113 requires nLockTime to be
strictly less than MTP, and MTP itself can reach but never exceed
UINT32_MAX (nTime is a uint32 field, and a block's timestamp must
exceed the MTP of the blocks preceding it, so once MTP hits UINT32_MAX
no further block -- and thus no further MTP advance -- is possible).
This is a general BIP113 property independent of OP_CLTV, so it lives
alongside the other BIP113 tests rather than in feature_cltv.py.

Christewart commented:
@sedited I believe all your points are integrated into https://github.com/bitcoin/bitcoin/pull/32229/commits/1f05d8e0364fa4b009e558c97e821be70313b8bc
